internal/provider/resource_managed_identities.go

```diff
@@ -0,0 +1,433 @@
+			Description: plan.Description.ValueString(),
+			Data:        encodedData,
+		})
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Error updating managed identity",
+			err.Error(),
+		)
+		return
+	}
+
+	// Copy all fields returned by Tharsis back into the plan.
+	copyManagedIdentity(*updated, &plan)
+
+	// Set the response state to the fully-populated plan.
+	resp.Diagnostics.Append(resp.State.Set(ctx, plan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+}
+
+func (t *managedIdentityResource) Delete(ctx context.Context,
+	req resource.DeleteRequest, resp *resource.DeleteResponse) {
+
+	// Get the current state.
+	var state ManagedIdentityModel
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Delete the managed identity via Tharsis.
+	// The ID is used to find the record to delete.
+	err := t.client.ManagedIdentity.DeleteManagedIdentity(ctx,
+		&ttypes.DeleteManagedIdentityInput{
+			ID: state.ID.ValueString(),
+		})
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Error deleting managed identity",
+			err.Error(),
+		)
+		return
+	}
+}
+
+// ImportState helps the provider implement the ResourceWithImportState interface.
+func (t *managedIdentityResource) ImportState(ctx context.Context,
+	req resource.ImportStateRequest, resp *resource.ImportStateResponse) {
+
+	// Retrieve import ID and save to id attribute
+	resource.ImportStatePassthroughID(ctx, path.Root("id"), req, resp)
+}
+
+// copyManagedIdentity copies the contents of a managed identity.
+// It is intended to copy from a struct returned by Tharsis to a Terraform plan or state.
+func copyManagedIdentity(src ttypes.ManagedIdentity, dest *ManagedIdentityModel) error {
+
+	decodedData, err := decodeDataString(src.Data)
+	if err != nil {
+		return err
+	}
+
+	dest.ID = types.StringValue(src.Metadata.ID)
+	dest.Type = types.StringValue(string(src.Type))
+	dest.ResourcePath = types.StringValue(src.ResourcePath)
+	dest.Name = types.StringValue(src.Name)
+	dest.Description = types.StringValue(src.Description)
+	dest.GroupPath = types.StringValue(getGroupPath(src.ResourcePath))
+	if decodedData.AWSRole != nil {
+		dest.AWSRole = types.StringValue(*decodedData.AWSRole)
+	}
+	if decodedData.AzureClientID != nil {
+		dest.AzureClientID = types.StringValue(*decodedData.AzureClientID)
+	}
+	if decodedData.AzureTenantID != nil {
+		dest.AzureTenantID = types.StringValue(*decodedData.AzureTenantID)
+	}
+	dest.Subject = types.StringValue(decodedData.Subject)
+
+	// Must use time value from SDK/API.  Using time.Now() is not reliable.
+	dest.LastUpdated = types.StringValue(src.Metadata.LastUpdatedTimestamp.Format(time.RFC850))
+
+	return nil
+}
+
+// encodeDataString checks the AWS role, Azure client ID, Azure tenant ID, and subject fields
+// and then marshals them into the appropriate type and base64 encodes that.
+func encodeDataString(managedIdentityType types.String, input managedIdentityDataInput) (string, error) {
```

Comment:
These helper functions should be added to the resource struct to prevent adding them to the global package namespace.